libs/@local/hashql/compiletest/src/suite/hir_lower_inference.rs

```diff
@@ -0,0 +1,130 @@
+pub(crate) struct HirLowerInferenceSuite;
+
+impl Suite for HirLowerInferenceSuite {
+    fn name(&self) -> &'static str {
+        "hir/lower/inference"
+    }
+
+    fn run<'heap>(
+        &self,
+        heap: &'heap Heap,
+        mut expr: Expr<'heap>,
+        diagnostics: &mut Vec<SuiteDiagnostic>,
+    ) -> Result<String, SuiteDiagnostic> {
+        let mut environment = Environment::new(expr.span, heap);
+        let registry = ModuleRegistry::new(&environment);
+
+        let (types, lower_diagnostics) = lower(
+            heap.intern_symbol("::main"),
+            &mut expr,
+            &environment,
+            &registry,
+        );
+
+        process_diagnostics(diagnostics, lower_diagnostics)?;
+
+        let interner = Interner::new(heap);
+        let (node, reify_diagnostics) = Node::from_ast(expr, &environment, &interner, &types);
+        process_diagnostics(diagnostics, reify_diagnostics)?;
+
+        let node = node.expect("should be `Some` if there are non-fatal errors");
+
+        let mut output = node
+            .pretty_print(&environment, PrettyOptions::default().without_color())
+            .to_string();
+
+        output.push_str("\n\n--------------------------------------\n\n");
+
+        let mut replacement = AliasReplacement::new(&interner);
+        let Ok(node) = replacement.fold_node(node);
+
+        let mut converter =
+            ConvertTypeConstructor::new(&interner, &types.locals, &registry, &environment);
+
+        let node = match converter.fold_node(node) {
+            Ok(node) => node,
+            Err(reported) => {
+                let diagnostic = process_diagnostics(diagnostics, reported)
+                    .expect_err("reported diagnostics should always be fatal");
+                return Err(diagnostic);
+            }
+        };
+
+        let mut inference = TypeInference::new(&environment, &registry);
+        inference.visit_node(&node);
+
+        // We sort so that the output is deterministic
+        let mut types: Vec<_> = inference
+            .types()
+            .iter()
+            .map(|(&hir_id, &type_id)| (hir_id, type_id))
+            .collect();
+        types.sort_unstable_by_key(|&(hir_id, _)| hir_id);
+
+        let (solver, inference_diagnostics) = inference.finish();
+        process_diagnostics(diagnostics, inference_diagnostics)?;
+
+        let (substitution, solver_diagnostics) = solver.solve();
+        process_diagnostics(diagnostics, solver_diagnostics.into_vec())?;
+
+        environment.substitution = substitution;
+
+        write!(
+            &mut output,
+            "{}",
+            node.pretty_print(&environment, PrettyOptions::default().without_color())
+        )
+        .expect("infallible");
+
+        for (hir_id, type_id) in types {
+            output.push_str("\n\n--------------------------------------\n\n");
+
+            output.push_str("Node:\n");
+
+            writeln!(
+                &mut output,
+                "{}\n\n",
```

Comment:
Quite a few newlines (noticeable in the `.stdout` files). No objections but I think a single `\n` is enough.

Reply:
I found that a single line makes readability hard when looking at complex types and expressions. I have on my backlog (will be creating a linear ticket tomorrow) to rework the output to be a bit more pleasant to review and look through.

Reply:
fair enough, the output is readable and that's the main thing 🙂